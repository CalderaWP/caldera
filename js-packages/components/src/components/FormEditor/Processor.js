import React, {Fragment, Component} from 'react';
import {HorizontalForm} from './HorizontalForm';
import {TabPanel} from '@wordpress/components';
import { Row,Column } from '@calderawp/factory';

import PropTypes from 'prop-types';

export class Processor extends Component {


	state = {
		activeTab: 'settings',
	};

	onSetTab = (activeTab) => {
		this.setState({activeTab});
	};

	render() {
		const {onClose, onRemove} = this.props;
		return (
			<Fragment>
				<Row>
					<TabPanel
						className="caldera-processor"
						activeClass="active-tab"
						onSelect={this.onSetTab}
						initialTabName={'settings'}
						tabs={[
							{
								name: 'settings',
								title: 'Settings',
								className: 'caldera-processor-settings-tab-btn',
							},
							{
								name: 'conditionals',
								title: 'Conditionals',
								className: 'caldera-processor-conditionals-tab-btn',
							},
						]}>
						{
							(tab) => {
								const {name} = tab;
								if( 'settings' === name ){
									return <HorizontalForm {...this.props} className={'caldera-processor-settings'}/>
								}
								return <div className={'caldera-processor-conditionals'}>Conditionals</div>
							}
						}
					</TabPanel>
				</Row>
				<Row>
					<Column
						width={'1/2'}
					>
						<button
							onClick={onClose}
						>
							Close
						</button>
					</Column>
					<Column
						width={'1/2'}
					>
						<button
							onClick={onRemove}
						>
							Remove
						</button>
					</Column>
				</Row>
			</Fragment>
		)
	}
}

Processor.propTypes = {
	...HorizontalForm.propTypes,
	initialActiveTab: PropTypes.string,
	onRemove: PropTypes.func,
};
Processor.defaultProps = {
	...HorizontalForm.defaultProps,
	initialActiveTab: 'settings'
};
