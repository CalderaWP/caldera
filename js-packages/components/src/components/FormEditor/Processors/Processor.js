import React, { Fragment, Component } from 'react';
import { HorizontalForm } from '../../HorizontalForm/HorizontalForm';
import { TabPanel } from '@wordpress/components';
import { Row, Column } from '@calderawp/factory';

import PropTypes from 'prop-types';

export class Processor extends Component {
	state = {
		activeTab: 'settings'
	};

	onSetTab = activeTab => {
		this.setState({ activeTab });
	};

	tabs = [
		{
			name: 'settings',
			title: 'Settings',
			className: 'caldera-processor-settings-tab-btn'
		},
		{
			name: 'conditionals',
			title: 'Conditionals',
			className:
				'caldera-processor-conditionals-tab-btn'
		}
	];
	render() {
		const { onClose, onRemove } = this.props;
		return (
			<Fragment>
				<Row>
					<TabPanel
						className="caldera-processor"
						activeClass="active-tab"
						onSelect={this.onSetTab}
						initialTabName={'settings'}
						tabs={this.tabs}
					>
						{() => {
							const {activeTab} = this.state;
							const tab =  this.tabs.find( t => t.name === activeTab );
							const { name } = tab;
							if ('settings' === name) {
								return (
									<HorizontalForm
										{...this.props}
										className={'caldera-processor-settings'}
									/>
								);
							}
							return (
								<div
									className={'caldera-processor-conditionals'}
								>
									Conditionals
								</div>
							);
						}}
					</TabPanel>
				</Row>
				<Row>
					<Column width={'1/2'}>
						<button
							className={'caldera-processor-close'}
							onClick={onClose}>
							Close
						</button>
					</Column>
					<Column width={'1/2'}>
						<button
							className={'caldera-processor-remove'}
							onClick={onRemove}>
							Remove
						</button>
					</Column>
				</Row>
			</Fragment>
		);
	}
}

//should have label!
Processor.propTypes = {
	...HorizontalForm.propTypes,
	initialActiveTab: PropTypes.string,
	onRemove: PropTypes.func,
	type: PropTypes.string,
	id: PropTypes.string,
};

Processor.defaultProps = {
	...HorizontalForm.defaultProps,
	initialActiveTab: 'settings'
};
