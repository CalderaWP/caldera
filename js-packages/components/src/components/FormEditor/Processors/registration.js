

const defaultRegistry = [];

/**
 * Add a processor type to form editor
 *
 * @param {*} processorType Defintion of processor type
 * @param {Array} registry Optional. Collection to add to -- generally you leave this at default.
 * @return {Array}
 */
export const registerProcessorType = (processorType,registry = defaultRegistry ) => {
	if( 'object' === typeof  processorType && processorType.hasOwnProperty('type' ) ) {
		return [ processorType,...registry];
	}

	return registry;

};

export const unregisterProcessorType = (processorType,registry = defaultRegistry ) => {
	if( 'object' === typeof  processorType && processorType.hasOwnProperty('type' ) ) {
		processorType = processorType.type;
	}
	const index = registry.findIndex( p => p.type === processorType );
	if( 0<= index ){
		registry.splice(index,1);
		return [ ...registry];
	}
	return registry;
};
