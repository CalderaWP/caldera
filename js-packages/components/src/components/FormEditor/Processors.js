import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import {Processor} from "./Processor";

export class Processors extends Component {

	state = {
		activeProcessorId: ''
	};

	/**
	 * Set which processor is active
	 * @param activeProcessorId
	 */
	setActive = (activeProcessorId) => {
		this.setState({activeProcessorId});
	};

	/**
	 * Check if a processor is active by ID
	 * @param processorId
	 * @return {string|boolean}
	 */
	isActiveProcessor = (processorId) => {
		const {activeProcessorId} = this.state;
		return activeProcessorId && processorId === activeProcessorId;
	};

	/**
	 * When a change in value for a processor is received, send all processors up.
	 *
	 * @param processorId
	 * @param updatedProcessor
	 */
	handleProcessorChange = (processorId, updatedProcessor) => {
		const {processors, updateProcessors} = this.props;
		let processorIndex = processors.findIndex(processor => {
			return processorId === processor.id
		});
		if (-1 !== processorIndex) {
			let processor = processors[processorIndex];
			let update = [
				...processors.splice(processorIndex - 1, 1),
				{
					...processor,
					...updatedProcessor
				}
			];
			updateProcessors(update);
		}


	};

	/**
	 * When a processor is removed, send updated list to change handler.
	 *
	 * @param processorId
	 */
	handleRemoveProcessor = (processorId) => {
		const {processors, updateProcessors} = this.props;
		updateProcessors([...processors.filter(processor => {
			return processorId !== processor.id
		})]);
	};

	/** @inheritDoc **/
	render() {
		const {activeProcessor} = this.state;
		const {processors} = this.props;
		return (
			<div>
				<div>
					{processors.map(processor => {
						const {
							id,
							fields,
							config,
							type,
							label
						} = processor;
						if (this.isActiveProcessor(id)) {

							return (
								<Fragment key={id}>
									<Processor
										className={`caldera-forms-active-processor-${id}`}
										fields={fields}
										initialValues={config}
										label={label}
										type={type}
										onChange={(fieldValues) => this.handleProcessorChange(id, fieldValues)}
										onRemove={() => this.handleRemoveProcessor(id)}
										onClose={() => this.setActive('')}
									/>
								</Fragment>
							)
						}
						return (
							<Fragment>
								<button
									className={`caldera-forms-choose-processor caldera-forms-choose-processor-${id}`}
									onClick={() => this.setActive(id)}
								>
									{label ? label : type}
								</button>
							</Fragment>

						)
					})}
				</div>
			</div>
		);
	}
}

export const processorsCollectionPropType = PropTypes.arrayOf(
	PropTypes.shape({
		id: PropTypes.string,
		label: PropTypes.string,
		type: PropTypes.string,
		fields: PropTypes.array,
		config: PropTypes.object
	})
);

Processors.propTypes = {
	processors: processorsCollectionPropType,
	updateProcessors: PropTypes.func,
	form: PropTypes.object,
	formFields: PropTypes.object
};

Processors.defaultProps = {
	processors: []
};
