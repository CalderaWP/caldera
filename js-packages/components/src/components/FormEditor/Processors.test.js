import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {Processor} from "./Processor";
import {Processors} from "./Processors";
import {mount} from 'enzyme';
import renderer from 'react-test-renderer';

import {
	checkboxFieldset,
	selectField,
	checkboxField,
	numberField,
	textField,
	emailField,
	radioField
} from '@calderawp/factory';

describe('Processors', () => {
	const processors = [
		{id: 'p1', type: 'Redirect', fields: []},
		{id: 'p2', type: 'Redirect', fields: []},
		{id: 'p3', type: 'Redirect', fields: []},
	];

	const form = {};
	const formFields = {};
	let updateProcessors;
	beforeEach(() => {
		updateProcessors = jest.fn();
	});


	it('Should update active', () => {
		const component = mount(
			<Processors
				processors={processors}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);
		component.instance().setActive('p2');
		expect(component.state('activeProcessorId')).toEqual('p2');
	});

	it('Can remove a processor', () => {
		let values = {};
		const updateProcessors = (updateValues) => {
			values = updateValues;
		};
		const component = mount(
			<Processors
				processors={processors}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);
		component.instance().handleRemoveProcessor('p2');

		expect(values.length).toEqual(2);
	});

	it( 'Can update a processor type', () => {
		const processors = [
			{id: 'p1', type: 'Redirect', fields: []},
			{id: 'p2', type: 'Redirect', fields: [],label: 'fLabel'},
		];
		let values = {};
		const updateProcessors = (updateValues) => {
			values = updateValues;
		};

		const component = mount(
			<Processors
				processors={processors}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);

		component.instance().handleProcessorChange('p2', {
			type: 'strident'
		});
		expect(values.find( p => 'p2' === p.id).label).toEqual('fLabel');
		expect(values.find( p => 'p2' === p.id).type).toEqual('strident');


	});


	it( 'Renders with a list of processors', () => {
		const processors = [{id: 'p1', type: 'Redirect', fields: [checkboxField,selectField]}];
		const component = renderer.create(
			<Processors
				processors={processors}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);
		expect(component.toJSON()).toMatchSnapshot();
	});


});
