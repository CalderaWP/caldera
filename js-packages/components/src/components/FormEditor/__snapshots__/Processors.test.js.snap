// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Processors collection in processors UI Renders with a list of processors 1`] = `
<div>
  <div>
    <div
      className="sc-bdVaJa caldera-row sc-bwzfXH pjOvZ"
      id={undefined}
    >
      <button
        className="caldera-forms-choose-processor caldera-forms-choose-processor-season7Episode2"
        onClick={[Function]}
      >
        Test sending form data to test API
      </button>
    </div>
    <div
      className="sc-bdVaJa caldera-row sc-bwzfXH pjOvZ"
      id={undefined}
    >
      <button
        className="caldera-forms-choose-processor caldera-forms-choose-processor-p2"
        onClick={[Function]}
      >
        Main MailChimp
      </button>
    </div>
  </div>
</div>
`;
